### PR TITLE
refactor(connlib): use events to handle ICE candidates

### PR DESCRIPTION
**Pull request by thomaseizinger**
Instead of using the `ControlSignal` trait for ICE candidates, we introduce a `poll_next_event` function on `Tunnel` that reads ICE candidates from the various receivers connected to the callbacks of the `PeerConnection`.

There are still a few TODOs in the code and it relies on unpublished code in `futures-bounded` but I am putting it up for early review and to see whether CI passes.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
@conectado Would appreciate an early review of this.

File: rust/connlib/tunnel/src/control_protocol.rs
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/control_protocol.rs]: I couldn't quite tell whether we actually need this two-phase mechanism here where we don't immediately start reading ICE candidates from the channel. If it is not needed, we could remove this function and immediately send ICE candidates to the other party.

File: rust/connlib/clients/shared/src/control.rs
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/control.rs]: Needs this on both sides is kind of unfortunate. I think we could probably solve that quite elegantly if `Tunnel` had a type parameter of the "remote ID". In other words, a tunnel on the gateway is parameterized with `ClientId` and vice versa. That would make these things type-safe.

File: rust/connlib/tunnel/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/lib.rs]: TODO: Decide on good values here.

File: rust/connlib/tunnel/src/control_protocol.rs
Comment (conectado) [reply rust/connlib/tunnel/src/control_protocol.rs]: The reason for this mechanism is that we don't want to send ICE candidates before we receive an SDP from the gateway since that could lead to the gateway receiving candidates to an unknown channel, we could buffer them in the gateway-side instead of the client-side but I found this easier to implement.

Comment (conectado) [reply rust/connlib/tunnel/src/control_protocol.rs]: In the gateway, however, it's fine to immediately start sending, it's done like that just to increase code sharing.

File: rust/connlib/clients/shared/src/control.rs
Comment (conectado) [reply rust/connlib/clients/shared/src/control.rs]: We will extract the `IceCandidateGatherer` from `Tunnel` when we replace `webrtc`, right? we can do it then.

File: rust/connlib/clients/shared/src/control.rs
Comment (conectado) [inline rust/connlib/clients/shared/src/control.rs]: We can just log the error.

If the connection can't succeed due to this, the connection will be re-attempted by an upper layer.

**Review by conectado — COMMENTED** [review]
Overall I think this is the correct approach until we have our own way to gather ICE candidates that doesn't depend on callbacks.

File: rust/connlib/clients/shared/src/control.rs
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/control.rs]: On a meta note, I am bit worried to fall for the [second-system effect](https://en.wikipedia.org/wiki/Second-system_effect) here where we think "we can fix all these things once we do X".

I think it might be worth it to explore before already, how we can use the type system effectively to help us with using `Tunnel` in its two configurations: on the client side and on the gateway side.

**Comment by thomaseizinger** [comment]
@conectado I had an attempt at making it more type-safe and it actually turned out quite well I think. Due to the existing split of functions on the `Tunnel` into different modules, I was able to "specialise" some of these impls to not be generic over the `IceState` but specify a particular one.

This allows us to directly call functions on the two different ICE states and only be generic over the final `poll` function which is now type-safe towards the kind of ID we are returning.

Curious to hear what you think!

**Comment by conectado** [comment]
> @conectado I had an attempt at making it more type-safe and it actually turned out quite well I think. Due to the existing split of functions on the `Tunnel` into different modules, I was able to "specialise" some of these impls to not be generic over the `IceState` but specify a particular one.
> 
> This allows us to directly call functions on the two different ICE states and only be generic over the final `poll` function which is now type-safe towards the kind of ID we are returning.
> 
> Curious to hear what you think!

Looks great but while we move to the new architecture I'd be wary of adding too many generic parameters on `Tunnel`, one extra is okay but I've had project where it grows to 10+ parameters and it starts being tedious to write the type and quite confusing to read!

File: rust/connlib/tunnel/src/lib.rs
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: Where are these used?

File: rust/connlib/tunnel/src/ice.rs
Comment (conectado) [inline rust/connlib/tunnel/src/ice.rs]: Perhaps we should move these to their own crate(client/gateway) to be clearer on where they are used, even if that means including a new dependency on those crates?

File: rust/connlib/tunnel/src/lib.rs
Comment (conectado) [inline rust/connlib/tunnel/src/lib.rs]: For events generated from the handler loops(such as `connection_intent` from `iface_handler` or now I'm working on a dns fallback), do you expect to follow a same pattern? we poll here for different `StreamMap`s?

**Review by conectado — APPROVED** [review]
I think this works! Left some non-blocking comments/questions

**Comment by thomaseizinger** [comment]
> > @conectado I had an attempt at making it more type-safe and it actually turned out quite well I think. Due to the existing split of functions on the `Tunnel` into different modules, I was able to "specialise" some of these impls to not be generic over the `IceState` but specify a particular one.
> > This allows us to directly call functions on the two different ICE states and only be generic over the final `poll` function which is now type-safe towards the kind of ID we are returning.
> > Curious to hear what you think!
> 
> Looks great but while we move to the new architecture I'd be wary of adding too many generic parameters on `Tunnel`, one extra is okay but I've had project where it grows to 10+ parameters and it starts being tedious to write the type and quite confusing to read!

Yeah I fully agree. We can always re-purpose the current `IceState` to a `role_state` and store all things in there that are specific to a particular role. This actually gives me an idea that on what I want to change here to make this clearer.

File: rust/connlib/tunnel/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/lib.rs]: Nowhere anymore! Will delete.

File: rust/connlib/tunnel/src/ice.rs
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/ice.rs]: Well, we can't have circular dependencies between tunnel and client/gateway so I think for the moment, we have to keep these in the `firezone-tunnel` crate. But yeah, in an ideal world, these would live outside. I think the way to do this is to move all the specific state _into_ these structs. What I'd like to avoid is adding more functions to `RoleState` to access specific things.

I'll document this on the trait.

File: rust/connlib/tunnel/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/lib.rs]: See my latest changes. I think the way we want to go is to add more stuff to the different role-states and emit the events accordingly. We should only modify the state where types are known though. That will help us separate functions such that they are not seemingly generic whilst they are only used in one place.

File: rust/connlib/tunnel/Cargo.toml
Comment (thomaseizinger) [inline rust/connlib/tunnel/Cargo.toml:28]: @conectado Are you okay with a git-dependency here for now or would you like to wait for a release?

I've proposed the changes in https://github.com/libp2p/rust-libp2p/pull/4616 but it might take a few days before this is landed.

File: rust/connlib/tunnel/src/control_protocol/client.rs
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/control_protocol/client.rs:287]: This was defined in the wrong module and the compiler told me once I've added the new state type parameters! :partying_face:

File: rust/connlib/tunnel/src/role_state.rs
Comment (conectado) [inline rust/connlib/tunnel/src/role_state.rs:26]: We should coordinate this with `webrtc`

File: rust/connlib/tunnel/Cargo.toml
Comment (conectado) [reply rust/connlib/tunnel/Cargo.toml:28]: I don't mind the git-dependency

File: rust/connlib/tunnel/src/control_protocol/client.rs
Comment (conectado) [reply rust/connlib/tunnel/src/control_protocol/client.rs:287]: nice!

**Review by conectado — APPROVED** [review]

File: rust/connlib/tunnel/src/role_state.rs
Comment (thomaseizinger) [reply rust/connlib/tunnel/src/role_state.rs:26]: Are you happy to just leave it for now? I think 5 minutes is pretty insane and for a good UX, this should be a lot less. I think any successful connection will be done before that and only if something goes wrong, we make sure that this is cleaned up.

Comment (conectado) [reply rust/connlib/tunnel/src/role_state.rs:26]: Oh yeah, I'd just leave a todo here or an explanation that there's an independent timeout in webrtc that will always timeout before this.

File: rust/connlib/tunnel/src/role_state.rs
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/role_state.rs:23]:
```suggestion
/// For how long we will attempt to gather ICE candidates before aborting.
///
/// Chosen arbitrarily.
/// Very likely, the actual WebRTC connection will timeout before this.
/// This timeout is just here to eventually clean-up tasks if they are somehow broken.
```

**Pull request merged**
